Kotlin Multiplatform Project Support (mpp)

drieks opened the pull request:
Hi @ftomassetti,

I added Kotlin Multiplatform Project Support (mpp), including an example.

The easiest way was to remove all configuration "magic" from the plugin. The Plugin is no longer needed, the jar now contains only the AntlrKotlinTask.

Both examples (for jvm and mpp) shows how to configure the task.

Kotlin was updated to 1.3.50, the Plugin Version to 0.0.5.

ftomassetti commented:
Thank you @drieks ! Your work is appreciated. However  the build seems to be failing

drieks commented:
I fixed the build and merged the new upstream changes into my PR.

drieks commented:
Hi @phyrian,

as I see, we worked both on MPP support :-)

I removed the "magic" done by the Gradle Plugin and added an example how to configure the Task manually. Please have a look at these files:

https://github.com/drieks/antlr-kotlin/blob/master/antlr-kotlin-examples-jvm/build.gradle.kts#L31-L55
https://github.com/drieks/antlr-kotlin/blob/master/antlr-kotlin-examples-mpp/build.gradle.kts#L65-L89

ftomassetti commented on .travis.yml:
```diff
@@ -6,4 +6,5 @@ script:
   - ./gradlew test
   - ./gradlew publishToMavenLocal
   - cd antlr-kotlin-examples-jvm && ../gradlew --info check && cd ..
+  - cd antlr-kotlin-examples-mpp && ../gradlew --info check && cd ..
```
nice!

ftomassetti commented on antlr-kotlin-examples-jvm/build.gradle.kts:
```diff
@@ -1,3 +1,16 @@
+
+buildscript {
+    // we have to re-declare this here :-(
+    val antlrKotlinVersion = "0.0.5"
```
that is ok, as this is like it was a separate project, something a user of ANTLR Kotlin would use, so it makes sense they have to specify the version number of a dependency they are using

ftomassetti commented on antlr-kotlin-examples-jvm/build.gradle.kts:
```diff
@@ -9,53 +22,53 @@ repositories {
-*/
 
-tasks.generateKotlinGrammarSource {
+// in antlr-kotlin-plugin <0.0.5, the configuration was applied by the plugin.
```
Is this something we should improve in the future?

ftomassetti commented on antlr-kotlin-examples-mpp/gradle:
```diff
@@ -0,0 +1 @@
+../gradle
```
do we need also the `gradle` command? Normally I just use `gradlew` and `gradle` is the one installed on the machine globally

drieks replied:
Hi @tfomassetti,
this is just a symlink

ftomassetti commented on antlr-kotlin-examples-mpp/src/commonAntlr/antlr:
```diff
@@ -0,0 +1 @@
+../../../antlr-kotlin-examples-jvm/src/main/antlr
```
I do not understand this file and the two following

drieks replied:
Hi @tfomassetti,
they are just a symlinks

phyrian commented:
Hi @drieks!
Does it work with custom outputDirectory? As I've seen you also commented out the custom path, and added back the default one.
I found that it needs to have the "antlr" source directory put in a folder named "main" and then it all works smoothly, but I don't really know where this condition comes from.
Please share your thoughts about it. It looks like you are much more into plugins :)

drieks commented:
Hi @ftomassetti,

I think, it is better to configure the Task manually. You just need to copy the example. All possible options are listed there. It was very, very hard before to understand how to change a setting. Or just getting a list of all possible options was nearly impossible. Also, the JavaPlugin was always added to the Build. The reason, why MPP was not working, was the "magic" of the Plugin. The AntlrKotlinTask was already working before.

Maybe we can add a default Value for antlrClasspath, but seeing the dependencies in the build should not be too confusing for users of antlr-kotlin, I hope..

drieks commented:
Hi @phyrian !

> Does it work with custom outputDirectory? As I've seen you also commented out the custom path, and added back the default one.

Maybe the comment above can be improved :-) There are no longer any default values. You can choose any directory you want :-) If you don't supply an outputDirectory, the Task should fail. (But ok, I did not test this..)
